docs/rfc/rfc-005-authentication.md

```diff
@@ -0,0 +1,132 @@
+
+### 4. Extensible Credential Mechanisms
+   
+Currently, the Cosmos SDK primarily supports accounts backed by public/private key pairings. However, there is a growing
+need to design and implement an extensible system that can support abstracted credentials, which are not limited to
+public/private key pairings. This would enable the development of more advanced and flexible authentication systems that
+cater to a broader range of use cases and security requirements.
+
+To address these problems, the Cosmos SDK needs to develop an authentication system that is closely integrated with the
+proposed account system, providing a robust and extensible framework for managing authentication credentials and supporting
+a variety of credential mechanisms.
+
+## Implementation
+
+We have initially mentioned that `x/accounts` only provides for a way to deploy accounts, but they're not backed by any
+credential, which means no state transition on behalf of the account can be incepted from a TX.
+
+`x/authn` is meant to provide the link between accounts and their TX credentials.
+Specifically `x/authn` defines the following `tx` gRPC interface:
+
+```protobuf
+// Msg defines the Msg service.
+service Msg {
+  rpc CreateAuthenticatedAccount(MsgCreateAuthenticatedAccount) returns (MsgCreateAuthenticatedAccountResponse) {};
+  rpc UpdateCredentials(MsgUpdateCredentials) returns (MsgUpdateCredentialsResponse) {};
+  rpc DeleteCredentials(MsgDeleteCredentials) returns (MsgDeleteCredentials) {};
+}
+message MsgCreateAuthenticatedAccount {
+  string sender = 1;
+  google.Protobuf.Any credential = 2;
+  accounts.MsgDeploy deploy_msg = 3;
+}
+message MsgCreateAuthenticatedAccountResponse {
+  accounts.MsgDeployResponse deploy_response = 1;
+}
+message MsgUpdateCredentials {
+  string sender = 1;
+  string kind = 2;
+  google.Protobuf.Any new_credential = 3;
+}
+message MsgUpdateCredentialsResponse {}
+message MsgDeleteCredentials {
+  string sender = 1;
+}
+message MsgDeleteCredentialsResponse {}
+```
+
+### MsgCreateAuthenticatedAccount
+
+This message contains an opaque credential defined as `google.Protobuf.Any`, alongside an `accounts.MsgDeploy` request.
+This creates a new account and couples it with a credential.
+
+### MsgUpdateCredentials & MsgDeleteCredentials
+
+The former allows for credentials of an account to be updated the latter destroys credentials for an account, making it
+effectively impossible for the account to send state transitions from a TX forever, unless the account has logic to again
+update its credentials.
+
+### The credential interface
+
+The credential interface, represented in our gRPC `tx` interface as a `google.Protobuf.Any`, is implemented by any type
+which satisfies the following interface:
+
+```go
+package authn
+type Credential[T any, PT interface{ *T; proto.Message }] interface {
+	*T
+	VerifySignedBytes(msgBytes []byte, signature []byte) bool
+}
```

Comment:
I think we should be thinking about a design which moves us away from implementing golang interface methods on protobuf types. This has been discussed a number of times but specifically, it limits the possible backing implementations for signature verification. Maybe this is along the lines of what the accounts RFC is trying to address in terms of accounts being deployed code, not sure... but there should be some way that a credential (as an `Any`) is resolved to a credential handler which may either verify signed bytes or identify a module account. This would allow for novel signature algorithms to be deployed at runtime in contract code for instance.

Reply:
It's difficult to find a different design as things currently stand, in the end there's a part which you need to store in state and then an implementation part which is dependent on what is stored in state, you can decouple them for sure but you still need to create a relationship, somehow, between what is stored in state and how the component makes use of those bytes.

Reply:
Maybe we could abstract the interface to be `VerifySignature(credentialBytes, msg, sig []byte) (bool, error)` and then the implementation of credentials decides how to decode the bytes, this also assumes the credentials cannot change.